### PR TITLE
MAINT: Remove the unused keepdim argument from np.ufunc.accumulate

### PR DESCRIPTION
**Pull request by eric-wieser**

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -3842,10 +3842,13 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
     PyArray_Descr *otype = NULL;
     PyArrayObject *out = NULL;
     int keepdims = 0;
-    static char *kwlist1[] = {"array", "axis", "dtype",
-                                "out", "keepdims", NULL};
-    static char *kwlist2[] = {"array", "indices", "axis",
-                                "dtype", "out", NULL};
+    static char *reduce_kwlist[] = {
+        "array",            "axis", "dtype", "out", "keepdims", NULL};
```
Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Umm, looks like you may have a hard tab in here. I preferred to previous format to tell the truth, although the alignment was a bit of on the second line.

File: numpy/add_newdocs.py [hunk 2]
```diff
@@ -5919,6 +5919,13 @@ def luf(lamdaexpr, *args, **kwargs):
         First array
     B : array_like
         Second array
+    dtype : data-type code, optional
+        The type used to represent the intermediate results. Defaults
```
Comment (charris) [inline numpy/add_newdocs.py]: I don't think this is correct. I believe what happens is that the arrays are converted to the common type, or dtype if given, the operation done, and then converted to the `out` type if the `out` argument is given.

**Comment by charris** [comment]
Couple of style nits and probably the documentation is incorrect. The actual change looks OK.

File: numpy/add_newdocs.py (hunk 2 above)
Comment (charris) [inline numpy/add_newdocs.py]: You can check that that is the case by using various combinations of dtype and out.

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: That's a set of spaces to align columns, to make visually diffing the arg lists easier.

The previous format was just indented with a bunch of soft tabs until it roughly aligned. I definitely think that having all the list items on one line is an improvement, but could be persuaded to remove that alignment

Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Oh, and no hard tabs.

Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: There are no hard tabs here:

image

Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Multiple spaces after the `,`. It looks just awful.

File: numpy/add_newdocs.py (hunk 2 above)
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Totally missed the mark with this one - the arguments are just forwarded to the ufunc, including things like `where` which don't have all too sensible a result

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: You'd argue that that the alignment of argument names isn't worth the extra spacing then?

Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Yes ;) I understand the temptation, but it is a false god. Along the same lines, PEP8 disapproves of aligning on equals signs, etc.

Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: There might be some exceptions for tabular test data, where having data aligned in columns may be useful, but here is is just a distraction to (my) eye.

**Comment by eric-wieser** [comment]
Ok, should be good for another review

**Comment by charris** [comment]
OK, this looks good. However, current uses of `np.ufunc.accumulate` where the `keepdims` keyword is passed will error. The usual procedure is to issue a `FutureWarning` or `DeprecationWarning` warning for such cases and wait a couple of releases before making the change. I think `FutureWarning` is the way to go here. So I propose to mark this with a 1.14 milestone and ask you to submit another PR with the warning. Just check if the `keepdims` keyword has been set and issue a warning if so. I also think the documentation should be changed to show all the current keywords and the `keepdims` keyword documented as being dropped in a future release.

The temptation is to skip this process as it seems unlikely that the keyword is actually used by anyone, but it is best to follow procedure.

**Comment by charris** [comment]
I apologize for not noticing the problem earlier, it just slipped my mind.

**Comment by eric-wieser** [comment]
Ok, broken into three PRs:

 * #8197 - the tiny doc fix
 * #8198 - deprecation of the argument
 * This one - the above with an additional removal commit, for ease of merging in future

**Comment by eric-wieser** [comment]
Ok, this one is all set for 1.14

**Comment by charris** [comment]
OK, thanks.

**Comment by eric-wieser** [comment]
Rebased

**Comment by mhvk** [comment]
This would seem OK to merge now, so I'll go ahead and do that. Nice to get the code cleaner, even if it is by a little bit at the time!

**Pull request merged**